### PR TITLE
Fix div-zero error

### LINKED ISSUE #71: Divide by zero error is water weight calculation
An error arises in `shared_tools.get_weight_at_cell()`.

```yaml
out_dir: 'output'
seed: 42
Length: 3000.
Width: 6000.
dx: 20
verbose: 1
Np_water: 2000
Np_sed: 2000
f_bedload: 0.35
C0_percent: 0.15
SLR: 0.00000072
save_eta_figs: True
save_eta_grids: True
save_stage_grids: True
save_depth_grids: True
save_discharge_grids: True
save_velocity_grids: True
save_dt: 10
```

leads to:

```python
Traceback (most recent call last):
  File "script/stack.py", line 10, in <module>
    delta.update()
  File "/home/andrew/Dropbox/software/pyDeltaRCM_WMT/pyDeltaRCM/model.py", line 92, in update
    self.run_one_timestep()
  File "/home/andrew/Dropbox/software/pyDeltaRCM_WMT/pyDeltaRCM/deltaRCM_tools.py", line 52, in run_one_timestep
    self.run_water_iteration()
  File "/home/andrew/Dropbox/software/pyDeltaRCM_WMT/pyDeltaRCM/water_tools.py", line 44, in run_water_iteration
    self.get_water_weight_array()
  File "/home/andrew/Dropbox/software/pyDeltaRCM_WMT/pyDeltaRCM/water_tools.py", line 193, in get_water_weight_array
    self.water_weights[i, j] = shared_tools.get_weight_at_cell(
ZeroDivisionError: division by zero
```

### PR DESCRIPTION
**Pull request by elbeejay**
Should close #71. 

### Short Explanation

In the case of an (x, y) location in which all adjacent cells and the cell itself have a weight value of `nan`, then even though the [line](https://github.com/DeltaRCM/pyDeltaRCM/blob/dea1069cd70aa994c749d6a86f9533749e39e919/pyDeltaRCM/shared_tools.py#L196) in `shared_tools.get_weight_at_cell` is not assigning values to any indices, it is still being executed and trying to divide by 0 (as `len(weight[~nanWeight]) == 0`). So the proposed fix is to use `np.maximum()` to ensure that the value there is always 1 or greater. An alternative solution is extending the [if-else statement](https://github.com/DeltaRCM/pyDeltaRCM/blob/dea1069cd70aa994c749d6a86f9533749e39e919/pyDeltaRCM/shared_tools.py#L192) but I figured `np.maximum` was going to be cheaper computationally than another clause (to split the case of having a 0 and the case of all nans).

### Longer Explanation

Using the config described in #71, the error is a bit cryptic as the division by zero is not actually happening within `water_tools` but rather within a jitted function in `shared_tools`. The error is also a bit puzzling as it occurs on one of the 'land' cell indices (2, 142) but wasn't thrown for the previous index (2, 141) which is also a 'land' cell. To better understand this, some of the variable values that are fed to `shared_tools.get_weight_at_cell` are provided below for both index (2, 141) and (2, 142).

For index (2, 141):
```
stage_nbrs = array([[0.1152    , 0.12517995, 0.12517995],
       [0.1152    , 0.12517995, 0.12517995],
       [0.1152    , 0.12517995, 0.12517995]], dtype=float32)

depth_nbrs = array([[0.09119999, 0.10117994, 0.10117994],
       [0.0952    , 0.10517995, 0.10517995],
       [0.0992    , 0.10917994, 0.10917994]], dtype=float32)

ct_nbrs = array([[-2, -2, -2],
       [-2, -2, -2],
       [-2, -2, -2]])

self.stage[i,j] = 0.12517995

self.qx[i,j] = 0.0

self.qy[i,j] = 0.0
```

For index (2, 142)
```
stage_nbrs = array([[0.12517995, 0.12517995, 0.12517995],
       [0.12517995, 0.12517995, 0.12517995],
       [0.12517995, 0.12517995, 0.12026037]], dtype=float32)

depth_nbrs = array([[0.10117994, 0.10117994, 0.10117994],
       [0.10517995, 0.10517995, 0.10517995],
       [0.10917994, 0.10917994, 0.10426037]], dtype=float32)

ct_nbrs = array([[-2, -2, -2],
       [-2, -2, -2],
       [-2, -2, -2]])

self.stage[i,j] = 0.12517995

self.qx[i,j] = 0.0

self.qy[i,j] = 0.0
```

The significant difference boils down to the values in `depth_nbrs` which are acquired from the padded depth array in `water_tools.get_water_weight_array()`. Some of these values for index (2, 141) are actually below the `dry_depth` threshold of 0.1 while for index (2, 142), all `depth_nbrs` values are above 0.1. This ultimately leads to a difference within the `shared_tools.get_weight_at_cell()` function (after [this line](https://github.com/DeltaRCM/pyDeltaRCM/blob/dea1069cd70aa994c749d6a86f9533749e39e919/pyDeltaRCM/shared_tools.py#L188) is executed):

For index (2, 141):
```
weight[depth_nbrs <= dry_depth] = 0
array([ 0., nan, nan,  0., nan, nan,  0., nan, nan])
```

For index (2, 142):
```
weight[depth_nbrs <= dry_depth] = 0
array([nan, nan, nan, nan, nan, nan, nan, nan, nan])
```

Then when the final weight setting is being done in the if-else statement there is a division by zero issue for index (2, 142) that is not present for index (2, 141) because the weight array contains some 0 values. As we can see in the example above, this difference in 0 v `np.nan` values is stemming from the `depth_nbrs` v `dry_depth` relationship.

So while this small bugfix should alleviate the bug pointed out in #71, there is the question of whether this is the best way to handle the `np.nan` v 0 value in the weighting. There are likely other alternative formulations to get around this issue besides the one proposed here, and the alternative mentioned above involving the addition of another clause to the [if-else statement](https://github.com/DeltaRCM/pyDeltaRCM/blob/dea1069cd70aa994c749d6a86f9533749e39e919/pyDeltaRCM/shared_tools.py#L192), so of course I'm open to other solutions.

### CODE REVIEW

File: pyDeltaRCM/shared_tools.py
Comment (amoodie) [inline pyDeltaRCM/shared_tools.py:197]: amazing when what I'm sure was a few hours of cumulative effort boils down to a 15-character single-line change.

Your way of fixing is probably faster than the if-else tree I set up, and I see why taking the `maximum` prevents the div-by-zero error, but I just want to check that the new behavior is as expected.

If I understand correctly, if going into the if-else `weight = array([nan, nan, nan, nan, nan, nan, nan, nan, nan])`, then coming out `weight = array([0, 0, 0, 0, 0, 0, 0, 0, 0])`. 

Would this be problematic for anything? The expected behavior on the way the old maintainers wrote it suggests that `weight` should always come back with something.  E.g., `weight = array([0, 0, nan, nan, nan, nan, nan, nan, nan])`would become `weight = array([0.5, 0.5, 0, 0, 0, 0, 0, 0, 0])`. 

I'm guessing that we only started to have this problem based on the change Eric made to compute the weights for every cell only once at the top of the water parcel iteration, rather than with every step of the water parcel, because this change made it so that we actually calculate weights of "land" cells at all. To this end, I begun to look at just skipping the weight computation altogether for cells with type==land (returning some dummy weight), but never explored it really. 

I guess that your fix would probably work all the same though, and it probably doesn't make much difference for computational speed. 

Anyway, I'm approving these changes, but just wanted to run my ideas by you before merge.

**Review by amoodie — APPROVED** [review]

File: pyDeltaRCM/shared_tools.py
Comment (elbeejay) [reply pyDeltaRCM/shared_tools.py:197]: > I'm guessing that we only started to have this problem based on the change Eric made to compute the weights for every cell only once at the top of the water parcel iteration, rather than with every step of the water parcel, because this change made it so that we actually calculate weights of "land" cells at all.

This is my take on it as well. I think that the singular computation of weights that we are now doing across the entire array (even across -2 'land' cells) has saved us a lot of time, but a consequence was creating this edge case for the weights. So I agree with you, and yes the output would be `weight = array([0, 0, 0, 0, 0, 0, 0, 0, 0])` - this would only present problems if there was another bug allowing parcels to route into "invalid" cells where they'd encounter the fully 0 weight arrays. 

> I guess that your fix would probably work all the same though, and it probably doesn't make much difference for computational speed.

Yeah I thought about this a bit and figured that the land area is typically a small portion of the domain, so I couldn't imagine it was a big deal.

Comment (amoodie) [reply pyDeltaRCM/shared_tools.py:197]: cool I agree with all of this. Merging now.

**Pull request merged**
